test/subscription_map.hpp

```diff
@@ -38,28 +38,103 @@
  *           Using a published topic, we can find all topic filters which match the specified topic
  *       . A stored topic map, storing topic -> data
  *           Using a new topic filter, we can find all stored topics which match the specified topic filter
+ *
+ *   Subscription map stores all entries in a tree
+ *   the tree starts from a root node, and topic filters are tokenized and stored in the tree
+ *
+ *   For example if the topic_filter example/monitor/Clients is stored, the following nodes are created:
+ *   root -> example -> monitor -> Clients
+ *
+ *   Every node in the tree may store one or multiple subscribers. Nodes store a reference count to the number of subscribers
+ *   so for example, if we store the following topic_filters:
+ *     example/
+ *     example/monitor/Clients
+ *
+ *   the subscription map looks as follows:
+ *     root(2) -> example(2) -> monitor(1) -> Clients (1)
+ *
+ *   hash and + are stored as normal nodes within the tree, but the parent node knows if a hash child is available. This
+ *   improves the matching, no extra lookup is required to see if a # or + child is available in a child node:
+ *
+ *     example/#
+ *
+ *    stores the following tree:
+ *      root -> example (hash: yes) -> #
+ *
+ *    and
+ *
+ *      example/+
+ *
+ *    stores the following tree:
+ *      root -> example (plus: yes) -> #
+ *
+ *    all node entries are stored in a single hash map. The key for every node is: (parent node id, path)
+ *
+ *      so if we store: root/example/test
+ *      root (id:1) -> example (id:2, key:1,example) -> test (id:3, key:2,test)
+ *
+ *    also, every node stores the key of its parent, allowing quick traversing from leaf to root of the tree
  */
 
+// Combined storage for count and flags
+// we can have 32bit or 64bit version
+template<std::size_t N>
+struct count_storage
+{
+    // Make new implementation for unsupported targets
+};
```

Comment:
Do you think  empty definition is better than no definition ?
I think that the compile error that is caused by no definition is more straight forward message.
Empty definition also make compile error but it may "count_strage has no member named ...".

By ye way, I agree with 64 bit by default. I think that it is achieved at *1 implementation below.

Reply:
Ah yes, no definition would be better. But what do you mean by *1 ?

Reply:
*1 means the point of location https://github.com/redboltz/mqtt_cpp/pull/717#discussion_r516612691

Let me elaborate.
`using count_storage_t = count_storage<sizeof(void *)>;` is the only place using `count_storage`.
`count_storage_t` is automatically adjusted pointer size. So typical 64bit platform is automatically choose count_storege<8>`. Some of 32bit platform automatically choose `count_storage<4>`. Otherwise compile error.
I think it is  good.

Reply:
Note:
If you want to fallback 128 bit to 64 bit, you can use https://wandbox.org/permlink/vRTaUMeh1NJg49qi